src/utils/loadable.ts

```diff
@@ -0,0 +1,21 @@
+export interface LoadableComponent {
+  componentWillLoad: () => Promise<void> | void;
+
+  componentDidLoad: () => Promise<void> | void;
+}
+
+const resolveMap = new WeakMap<LoadableComponent, (value: void | PromiseLike<void>) => void>();
+
+const promiseMap = new WeakMap<LoadableComponent, Promise<void>>();
+
+export function connectLoadableComponent(component: LoadableComponent): void {
+  promiseMap.set(component, new Promise((resolve) => resolveMap.set(component, resolve)));
```

Comment:
Sidebar: we need a different naming scheme for these utils. I think we discussed this briefly before w/ @Elijbet on how we started using `connect`/`disconnect` to match the `connectedCallback`/`disconnectedCallback` lifecycle hooks, but this leads to disconnected naming (IMO).

Reply:
yeah. If we wanted to match the lifecycle the names might be weird. I guess we could revisit later?

Reply:
Yeah, we can revisit later. 

@Elijbet Can you create a follow-up issue with some possible ideas on how to name these helpers? Thanks! 🙇